sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala

```diff
@@ -0,0 +1,471 @@
+    assert(snapshotVersion.nonEmpty, "snapshot file not generated")
+
+    // After version 20, snapshotting should generate newer snapshot files
+    updateVersionTo(20)
+    require(getDataFromFiles(provider) === Set("a" -> 20), "store not updated correctly")
+    provider.manage()       // do snapshot
+    assert(getDataFromFiles(provider) === Set("a" -> 20), "snapshotting messed up the data")
+    assert(getDataFromFiles(provider) === Set("a" -> 20))
+
+    val latestSnapshotVersion = (0 to 20).filter(version =>
+      fileExists(provider, version, isSnapshot = true)).lastOption
+    assert(latestSnapshotVersion.nonEmpty, "no snapshot file found")
+    assert(latestSnapshotVersion.get > snapshotVersion.get, "newer snapshot not generated")
+
+  }
+
+  test("cleaning") {
+    val provider = newStoreProvider(maxDeltaChainForSnapshots = 5)
+
+    for (i <- 1 to 20) {
+      val store = provider.getStore(i - 1)
+      update(store, "a", i)
+      store.commit()
+      provider.manage() // do cleanup
+    }
+    require(
+      unwrapToSet(provider.latestIterator()) === Set("a" -> 20),
+      "store not updated correctly")
+
+    assert(!fileExists(provider, version = 1, isSnapshot = false)) // first file should be deleted
+
+    // last couple of versions should be retrievable
+    assert(getDataFromFiles(provider, 20) === Set("a" -> 20))
+    assert(getDataFromFiles(provider, 19) === Set("a" -> 19))
+  }
+
+  test("StateStore.get") {
+    quietly {
+      val dir = Utils.createDirectory(tempDir, Random.nextString(5)).toString
+      val storeId = StateStoreId(0, 0)
+
+      // Verify that trying to get incorrect versions throw errors
+      intercept[IllegalArgumentException] {
+        StateStore.get(storeId, dir, -1)
+      }
+      assert(!StateStore.isLoaded(storeId)) // version -1 should not attempt to load the store
+
+      intercept[IllegalStateException] {
+        StateStore.get(storeId, dir, 1)
+      }
+
+      // Increase version of the store
+      val store0 = StateStore.get(storeId, dir, 0)
+      assert(store0.version === 0)
+      update(store0, "a", 1)
+      store0.commit()
+
+      assert(StateStore.get(storeId, dir, 1).version == 1)
+      assert(StateStore.get(storeId, dir, 0).version == 0)
+
+      // Verify that you can remove the store and still reload and use it
+      StateStore.remove(storeId)
+      assert(!StateStore.isLoaded(storeId))
+
+      val store1 = StateStore.get(storeId, dir, 1)
+      assert(StateStore.isLoaded(storeId))
+      update(store1, "a", 2)
+      assert(store1.commit() === 2)
+      assert(unwrapToSet(store1.iterator()) === Set("a" -> 2))
+    }
+  }
+
+  test("background management") {
+    val conf = new SparkConf()
+      .setMaster("local")
+      .setAppName("test")
+      .set("spark.sql.streaming.stateStore.managementInterval", "10ms")
+    val opId = 0
+    val storeId = StateStoreId(opId, 0)
+    val dir = Utils.createDirectory(tempDir, Random.nextString(5)).toString
```

Comment:
I can use createTempDir here because i dont need to use it across SparkContexts. But the root tempDir is going to be deleted in afterAll() anyways.